test/cloudnative/agentversion_set.go

```diff
@@ -0,0 +1,191 @@
+	"github.com/stretchr/testify/require"
+	"golang.org/x/exp/rand"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/e2e-framework/klient/k8s"
+	"sigs.k8s.io/e2e-framework/klient/wait"
+	"sigs.k8s.io/e2e-framework/klient/wait/conditions"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/pkg/features"
+)
+
+const VERSION = "VERSION"
+
+var injectionLabel = map[string]string{
+	"inject": "dynakube",
+}
+
+func AgentVersionSet(t *testing.T) features.Feature {
+	secretConfig, err := secrets.DefaultSingleTenant(afero.NewOsFs())
+	require.NoError(t, err)
+
+	versions := getAvailableVersions(secretConfig, t)
+	oldVersion, newVersion := assignVersions(t, versions, version.SemanticVersion{}, version.SemanticVersion{})
+
+	dk := dynakube.NewBuilder().
+		WithDefaultObjectMeta().
+		WithDynakubeNamespaceSelector().
+		ApiUrl(secretConfig.ApiUrl).
+		CloudNativeWithAgentVersion(&v1beta1.CloudNativeFullStackSpec{}, oldVersion).
+		Build()
+
+	agentVersionset := features.New("cloudnative with agent version set")
+	agentVersionset.Setup(namespace.Create(
+		namespace.NewBuilder("test-namespace-1").
+			WithLabels(injectionLabel).
+			Build()),
+	)
+	agentVersionset.Setup(secrets.ApplyDefault(secretConfig))
+	agentVersionset.Setup(operator.InstallAllForKubernetes())
+	agentVersionset.Setup(manifests.InstallFromFile("../testdata/cloudnative/sample-deployment.yaml"))
+
+	setup.AssessDeployment(agentVersionset)
+
+	agentVersionset.Assess("install dynakube", dynakube.Apply(dk))
+
+	assessVersionChecks(agentVersionset, oldVersion)
+
+	agentVersionset.Assess("update dynakube", updateDynakube(newVersion))
+	agentVersionset.Assess("dynakube phase changes to 'Running'",
+		dynakube.WaitForDynakubePhase(dynakube.NewBuilder().WithDefaultObjectMeta().Build()))
+	setup.AssessDeployment(agentVersionset)
+
+	assessVersionChecks(agentVersionset, newVersion)
+
+	return agentVersionset.Feature()
+}
+
+func getAvailableVersions(secret secrets.Secret, t *testing.T) []string {
+	dtc, err := dtclient.NewClient(secret.ApiUrl, secret.ApiToken, secret.ApiToken)
+	require.NoError(t, err)
+	versions, err := dtc.GetAgentVersions(dtclient.OsUnix, dtclient.InstallerTypeDefault, arch.FlavorDefault, arch.Arch)
+	require.NoError(t, err)
+
+	return versions
+}
+
+func assessVersionChecks(builder *features.FeatureBuilder, version version.SemanticVersion) {
+	builder.Assess("restart csi driver", restartCSIDriver)
+	builder.Assess("start sample apps and injection", sampleapps.Restart)
+	builder.Assess("check init containers", checkInitContainers)
+	builder.Assess("check env vars of init container", checkVersionInSampleApp(version))
+}
+
+func assignVersions(t *testing.T, versions []string,
+	old version.SemanticVersion, new version.SemanticVersion) (version.SemanticVersion, version.SemanticVersion) {
+
+	lowerHalf := versions[0 : len(versions)/2]
+	upperHalf := versions[len(versions)/2+1 : len(versions)-1]
```

Comment:
If I understand correctly, this algorithm wont work for `len(versions) < 2`

Reply:
you are right,thank you for pointing out! :)